projects/flimnap/main.cc

```diff
@@ -65,10 +65,30 @@ namespace {
     constexpr float fixed_velocity = 5.0f;
 
     // pose calculation loop
-    constexpr systime_t pose_loop_time = US2ST(8333); // update pose at 120 Hz
+    constexpr systime_t pose_loop_period = US2ST(8333); // update pose at 120 Hz
 
     // dynamics loop
-    constexpr systime_t looptime = MS2ST(1); // 1 ms -> 1 kHz
+    constexpr systime_t dynamics_loop_period = MS2ST(1); // 1 ms -> 1 kHz
+
+    // Suspends the invoking thread until the system time arrives to the
+    // specified value.
+    // The system time is assumed to be between prev and time else the call is
+    // assumed to have been called outside the allowed time interval, in this
+    // case the thread yields the time slot and no sleep is performed.
+    systime_t chThdSleepUntilWindowedOrYield(systime_t prev, systime_t next) {
+        systime_t time;
+
+        chSysLock();
+        time = chVTGetSystemTimeX();
+        if (chVTIsTimeWithinX(time, prev, next)) {
+            chThdSleepS(next - time);
+        } else {
+            chSchDoYieldS();
+        }
+        chSysUnlock();
```

Comment:
The main reason is due to the ChibiOS system state machine: http://chibios.sourceforge.net/html/concepts.html#system_states

Certain functions can only be called in critical zones (S-Locked or I-Locked) those with *S or *I suffix to denote the respective locked state. In order to sleep a thread, the system __always__ needs to be locked. The function variant without the *S suffix `chThdSleep()`  actually locks and unlocks (`os/rt/src/chthreads.c`):
```cpp
void chThdSleep(systime_t time) {

  chSysLock();
  chThdSleepS(time);
  chSysUnlock();
}
```
which then in turn calls (`os/rt/src/chschd.c`):
```cpp
msg_t chSchGoSleepTimeoutS(tstate_t newstate, systime_t time) {

  chDbgCheckClassS();

  if (TIME_INFINITE != time) {
    virtual_timer_t vt;

    chVTDoSetI(&vt, time, wakeup, currp);
    chSchGoSleepS(newstate);
    if (chVTIsArmedI(&vt)) {
      chVTDoResetI(&vt);
    }
  }
  else {
    chSchGoSleepS(newstate);
  }

  return currp->p_u.rdymsg;
}
```
which is arming a virtual timer to wake up the thread after some period. You certainly don't want your thread with a sleep call to get preempted while it's in the process for getting ready for bed. That'd be like someone coming into your room and changing the alarm time, or something.